ipalib/certmapping.py

```diff
@@ -0,0 +1,337 @@
+        extensions = [rule.template for rule in syntax_rules
+                      if rule.is_extension]
+
+        return {'parameters': parameters, 'extensions': extensions}
+
+    def _prepare_syntax_rule(
+            self, syntax_rule, data_rules, description, data_sources):
+        """Overrides method to pull out whether rule is an extension or not."""
+        prepared_template = super(OpenSSLFormatter, self)._prepare_syntax_rule(
+            syntax_rule, data_rules, description, data_sources)
+        is_extension = syntax_rule.options.get('extension', False)
+        return self.SyntaxRule(prepared_template, is_extension)
+
+
+class CertutilFormatter(Formatter):
+    base_template_name = 'certutil_base.tmpl'
+
+    def _get_template_params(self, syntax_rules):
+        return {'options': syntax_rules}
+
+
+class FieldMapping(object):
+    """Representation of the rules needed to construct a complete cert field.
+
+    Attributes:
+        description: str, a name or description of this field, to be used in
+            messages
+        syntax_rule: Rule, the rule defining the syntax of this field
+        data_rules: list of Rule, the rules that produce data to be stored in
+            this field
+    """
+    __slots__ = ['description', 'syntax_rule', 'data_rules']
+
+    def __init__(self, description, syntax_rule, data_rules):
+        self.description = description
+        self.syntax_rule = syntax_rule
+        self.data_rules = data_rules
+
+
+class Rule(object):
+    __slots__ = ['name', 'template', 'options']
+
+    def __init__(self, name, template, options):
+        self.name = name
+        self.template = template
+        self.options = options
+
+
+class RuleProvider(object):
+    def rules_for_profile(self, profile_id, helper):
+        """
+        Return the rules needed to build a CSR for the given certificate
+        profile.
+
+        :param profile_id: str, name of the certificate profile to use
+        :param helper: str, name of tool (e.g. openssl, certutil) that will be
+            used to create CSR
+
+        :returns: list of FieldMapping, filled out with the appropriate rules
+        """
+        raise NotImplementedError('RuleProvider class must be subclassed')
+
+
+class FileRuleProvider(RuleProvider):
+    def __init__(self, csr_data_dir=CSR_DATA_DIR):
+        self.rules = {}
+        self.csr_data_dir = csr_data_dir
+
+    def _rule(self, rule_name, helper):
+        if (rule_name, helper) not in self.rules:
+            rule_path = os.path.join(self.csr_data_dir, 'rules',
+                                     '%s.json' % rule_name)
+            try:
+                with open(rule_path) as rule_file:
+                    ruleset = json.load(rule_file)
+            except IOError:
+                raise errors.CertificateMappingError(
+                    reason=_('Ruleset %(ruleset)s is configured in profile but'
+                             ' does not exist.') % {'ruleset': rule_name})
```

Comment:
This looks rather like errors.NotFound.

Also, this method has no knowledge of profiles, so how do you know the ruleset is configured in profile?